core/arch/arm/kernel/boot.c

```diff
@@ -1140,6 +1149,137 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
 				   CFG_TZDRAM_SIZE);
 }
 
+#ifdef CFG_OVERLAY_ADDR
+static int add_descriptor_node(struct dt_descriptor *dt, const char *driver)
+{
+	unsigned int i = 0, j = 0;
+	char root_name[256] = { '\0' };
+	char buffer[512] = { '\0' };
+	char *item[128] = { '\0' };
+	char *root = NULL;
+	char *p = buffer;
+	int offs = 0;
+
+	strncpy(buffer, driver, sizeof(buffer));
+
+	/* strok modifies the buffer */
+	root = strtok_r(p, "/", &p);
+	if (!root)
+		return -1;
+
+	for (i = 0; i < ARRAY_SIZE(item); i++) {
+		item[i] = strtok_r(p, "/", &p);
+		if (!item[i])
+			break;
+	}
+
+	if (i >= ARRAY_SIZE(item))
+		return -1;
+
+	if (fdt_path_offset(dt->blob, driver) >= 0)
+		return 0;
+
+	/* add the root node name */
+	snprintf(root_name, sizeof(root_name), "/%s", root);
+	offs = fdt_path_offset(dt->blob, root_name);
+	if (offs < 0) {
+		offs = add_dt_path_subnode(dt, "/", root);
+		if (offs < 0)
+			return -1;
+	}
+
+	for (j = 0; j < i; j++) {
+		offs = fdt_add_subnode(dt->blob, offs, item[j]);
+		if (offs < 0)
+			return -1;
+	}
+
+	return offs;
+}
+#endif
+
+/* Guarantee that these drivers will not be used by either EL1 or EL0: set the
+ * device tree node status in the FDT and overlay to disabled and the
+ * secure-status to okay.
+ */
+static int reserve_platform_drivers_dt_nodes(struct dt_descriptor *dt)
+{
+	static const char **drivers;
+	static size_t num_drivers;
+	static bool init;
+	const char * const *drvs = { NULL };
+	const char *path = NULL;
+	unsigned int i = 0;
+	int offs = 0;
+	int len = 0;
+
+#ifdef CFG_OVERLAY_ADDR
+	if (dt->is_overlay)
+		goto handle;
+#endif
+	if (init) {
+		/* OP-TEE handles first a DTB populated with a __symbols__ node
+		 * passed in parameters. Then, using those symbols, it generates
+		 * the overlay.
+		 */
+		goto handle;
+	}
+
+	offs = fdt_path_offset(dt->blob, "/__symbols__");
+	if (offs < 0)
+		goto handle;
+
+	drvs = main_get_optee_reserved_drivers(&num_drivers);
+	if (!drvs || !num_drivers)
+		return 0;
+
+	drivers = malloc(num_drivers * sizeof(drivers[0]));
+	if (!drivers)
+		return -1;
+
+	for (i = 0; i < num_drivers; i++) {
+		path = fdt_getprop_namelen(dt->blob, offs,
+					   drvs[i], strlen(drvs[i]), &len);
+		if (!path)
+			return -1;
+
+		drivers[i] = malloc(len);
+		if (!drivers[i])
+			return -1;
+
+		memcpy((void *)drivers[i], (void *)path, len);
+	}
+
+	init = true;
+
+handle:
+
+	for (i = 0; i < num_drivers; i++) {
+		if (!drivers[i])
+			break;
+#ifdef CFG_OVERLAY_ADDR
+		if (dt->is_overlay)
+			offs = add_descriptor_node(dt, drivers[i]);
+		else
+#endif
+			offs = fdt_path_offset(dt->blob, drivers[i]);
+
+		if (offs < 0)
+			return -1;
+
+		if (!offs)
```

Comment:
you are right. this is a bug. if it was the overlay it would have been ok (it meant it had already been reserved) But not for a DTB. will remove this check. thanks!